### ISSUE TITLE
[AVR] Implement `__mulsi3` in compiler-rt/builtins

### ISSUE THREAD

**Issue by benshi001** [llvm/llvm-project#55279]
`__mulsi3` is needed for int32 multiplication, and called by `__addsf3`/`__subsf3`/`__mulsf3`/`__divsf3`.

`__divmodsi4` is already implemented as common code.

**Comment by aykevl** [comment]
> `__mulsi3` is needed for int32 multiplication, and called by `__addsf3`/`__subsf3`/`__mulsf3`/`__divsf3`.

I have an implementation in Go that I can easily convert to C. Or do you have one already?

> `__divmodsi4` is already implemented as common code.

Unfortunately it is incorrect. The signature is different. AVR expects the following signature:

```c
struct {int32_t, int32_t} __divmodsi4(int32_t, int32_t);
```

It can be implemented this way (untested, not optimized):

```c
uint64_t  __udivmodsi4(uint32_t a, uint32_t b) {
    uint32_t div = __udivsi3(a, b);
    uint32_t rem = a - (div * b);
    return ((uint64_t)div) | (((uint64_t)rem)<<32);
}
```

**Comment by benshi001** [comment]
No, I have not implemented all int32 helper functions. I appreciate if you can submit your implementation of `__mulsi3`/`__divmodsi4`/`udivmodsi4`.

**Comment by benshi001** [comment]
As @aykevl mentioned in https://reviews.llvm.org/D123200

```
[@benshi001](https://reviews.llvm.org/p/benshi001/) I have been looking through the GCC code and I think avr-gcc also has a special calling convention for many other functions, including __mulqi3 and __mulhi3.

Source:

I think this is where the ABI is specified in the compiler: https://github.com/gcc-mirror/gcc/blob/releases/gcc-5.4.0/gcc/config/avr/avr.md#L1543-L1549 You can see that it multiplies R24 with R22 and stores the result in R24, and clobbers R22 in the process. But no other registers.
[In this code sample](https://godbolt.org/z/cPE5ajrbv), avr-gcc doesn't save char c (r20) across the __mulqi3 and __mulhi3 calls, which is normally call-clobbered.
Therefore, I think we need to be a bit more careful with defining these AVR builtins and check the ABI in avr-gcc first.
Also, we can make use of this and optimize the AVR backend more (you can see that the Clang generated code is much worse than avr-gcc in the above examples).
```

**Comment by benshi001** [comment]
I have checked with libgcc's code,

1. for (__mulhi3)[https://github.com/gcc-mirror/gcc/blob/releases/gcc-5.4.0/libgcc/config/avr/lib1funcs.S#L305], 
```
;;; R25:R24 = R23:R22 * R25:R24
;;; (C1:C0) = (A1:A0) * (B1:B0)
;;; Clobbers: __tmp_reg__, R21..R23
```

2. For (__mulqi3)[https://github.com/gcc-mirror/gcc/blob/releases/gcc-5.4.0/libgcc/config/avr/lib1funcs.S#L217]
```
r24 is used for return result, r22 and __tmp_reg__ are clobbered.
```

**Comment by benshi001** [comment]
I have refactored them at https://reviews.llvm.org/D125077

**Comment by aykevl** [comment]
I looked at `__mulsi3` today. There are two different versions:

1. For parts that support `mul`: there is a version with a special calling convention that only clobbers R26:R27, SREG, and R0.
2. For parts that don't support `mul`: these use a generic version that could be implemented in C but is probably more efficient to implement in assembly. I have one written already. Unfortunately, it does not adhere to the calling convention required for parts with `mul` support.

**Comment by aykevl** [comment]
I have implemented __mulsi3 here: https://reviews.llvm.org/D137581
It's slow (especially on chips with `mul` support) but it should work on all chips except avrtiny.

**Comment by aykevl** [comment]
And now for `__udivmodsi4` and `__divmodsi4`: https://reviews.llvm.org/D138166
